meshmode/array_context.py

```diff
@@ -437,77 +437,73 @@ def _get_array_container_ac_arith(ary: ArrayContainerWithArithmetic):
+    """A class decorator that makes the class to which it is applied a
+    ``frozen`` :func:`~dataclasses.dataclass` and registers appropriate
+    implementations of :func:`serialize_container` and :func:`deserialize_container`.
 
-    exec_dict = {}
-    exec(compile(code, "<generated code>", "exec"), exec_dict)
-    exec_dict["_MODULE_SOURCE_CODE"] = code
+    Attributes that are not array containers are allowed. In order to decide
+    whether an attribute is an array container, the declared attribute type
+    is checked for whether it is a subclass of :class:`ArrayContainer`.
+    Note that this classification does not necessarily agree with
+    :class:`is_array_container` (which is not usable here because it operates
+    on instances, not types). :mod:`numpy` object arrays are a particularly
+    important example of this. To document that an attribute *is* a
+    :mod:`numpy` object array, use :class:`NumpyObjectArray` in the
+    attribute type annotation.
+    """
+    from dataclasses import dataclass
+    cls = dataclass(frozen=True)(cls)
 
-    return exec_dict["_generated_serialize_dataclass"]
+    array_fields = [f for f in fields(cls)
+            if issubclass(f.type, _ARRAY_CONTAINER_LIKE)]
+    non_array_fields = [f for f in fields(cls)
+            if not issubclass(f.type, _ARRAY_CONTAINER_LIKE)]
 
+    if not array_fields:
+        raise ValueError(f"'{cls}' must have fields with array container type "
+                "in order to use the dataclass_array_container decorator")
 
-@memoize(key=lambda ary: (type(ary), _deserialize_dataclass_container))
-def _generate_dataclass_deserialization_func(template):
-    from pytools.codegen import remove_common_indentation
-    code = remove_common_indentation("""
-    def _generated_deserialize_dataclass(_template, _iterable):
-        kwargs = dict(_iterable)
-        return type(_template)({})
-    """.format(
-        ", ".join(
-            f"{f.name}=kwargs['{f.name}']"
-            if issubclass(f.type, (ArrayContainer, NumpyObjectArray))
-            else f"{f.name}=_template.{f.name}"
-            for f in fields(template))
-        )
-    )
+    ser_expr = ", ".join(f"({repr(f.name)}, ary.{f.name})" for f in array_fields)
+    template_kwargs = "".join(
+            f"{f.name}=template.{f.name}, " for f in non_array_fields)
 
-    exec_dict = {}
-    exec(compile(code, "<generated code>", "exec"), exec_dict)
-    exec_dict["_MODULE_SOURCE_CODE"] = code
+    from pytools.codegen import remove_common_indentation
+    ser_code = remove_common_indentation(f"""
+        from typing import Any, Iterable, Tuple
+        from meshmode.array_context import serialize_container, deserialize_container
 
-    return exec_dict["_generated_deserialize_dataclass"]
+        @serialize_container.register(cls)
+        def serialize_{cls.__name__}(ary: cls):
+            return ({ser_expr},)
 
+        @deserialize_container.register(cls)
+        def deserialize_{cls.__name__}(
```

Comment:
```suggestion
        def deserialize_{cls.__name__.lower()}(
```
? There should really be a `str.to_snake_case()` :(

Reply:
54ce090